### ISSUE TITLE
Rule for block/inline comments in SCSS

### ISSUE THREAD

**Issue by XOP** [stylelint/stylelint#866]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

I'd like to propose a rule for distinction comments' usage in SCSS.
In my guidelines, there is a basic principle, that separates block comments from inline.
Inline comments can only be used inside of braces (thus pointing to nesting and/or context),
and block comments used for all other purposes.
So this is correct:
```scss
/* comment */
.foo {
    color: #fff; // comment
}
```

And this is incorrect:
```scss
// comment
.bar {
    /* comment */
    color: #bbb;
}
```
Of course, this is only valid for **scss syntax**.

> Which rule, if any, is this issue related to?

There are some related rules:
`comment-empty-line-before`
`no-invalid-double-slash-comments`
`at-rule-empty-line-before`
`rule-non-nested-empty-line-before`
`rule-nested-empty-line-before`

> What CSS is needed to reproduce this issue?

here's all in places:
```scss
/* comment */
.foo {
    color: #fff; // comment
}
```

and there should be 2 errors:
- inline comment outside of braces
- block comment inside of braces
```scss
// comment
.bar {
    /* comment */
    color: #bbb;
}
```

> What stylelint configuration is needed to reproduce this issue?

Here's my config that is WIP currently,  
just because it needs some more rules to be accurate.
[https://www.npmjs.com/package/stylelint-config-xop](https://www.npmjs.com/package/stylelint-config-xop) 

> Which version of stylelint are you using?

`4.4.0`, will change to `4.5.0` since it has useful feature for disabling lines of rules and some extra.

> How are you running stylelint: CLI, PostCSS plugin, Node API?

I'm using Node API, 
CLI didn't work for me (different results on linting, but I guess this is _another issue_)

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

SCSS syntax, nesting, yes. And to most of CSS preprocessors, since they allow inline comments usage.

> What did you expect to happen?

errors / warnings as described above

> What actually happened (e.g. what warnings or errors you are getting)?

nothing, it's not implemented

So, finally. If I need this behaviour, should I go for plugin or rule is ok?
For me it looks more like combination of rules.
Something for disallowing usage of inline comments outside of braces etc.

**Comment by jeddy3** [comment]
Thanks for taking the time to write such a detailed issue :)

As an aside, your config is looking great. It's fantastic to see each of the rules you are using being individually tested!

> So, finally. If I need this behaviour, should I go for plugin or rule is ok?

This is the perfect situation for [building your own plugin](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md) as it's applicable to *non-standard* syntax.

> For me it looks more like combination of rules.

If it *is* a combination of rules, then you'll probably want to make use of [`stylelint.rules`](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintrules). `stylelint.rules` was added with this *exact* scenario in mind i.e. conditionally applying the standard rules within a plugin.

Having said that, I'm not sure it is a combination. As all the rules are *standalone* and I can't seen how any of the related rules you've listed will conflict with your proposed plugin i.e a plugin that checks for inline comment outside of braces and block comment inside of braces. 

> CLI didn't work for me (different results on linting, but I guess this is another issue)

If you have a moment can you please write (in as much details as possible) this up in another issue, so that we can investigate it?

I'm going see if anyone else has any input before marking this as a plugin and closing the issue.  

FYI, you can use [`stylelint-selector-bem-pattern`](https://github.com/davidtheclark/stylelint-selector-bem-pattern) as a reference for building a plugin if you like.

**Comment by davidtheclark** [comment]
Yes, I agree that this is a candidate for a plugin, because it is about preprocessor-specific syntax.

@kristerkari was thinking about starting a collection of SCSS-specific plugins over at https://github.com/kristerkari/stylelint-scss; maybe you want to talk to him about your idea?

**Comment by XOP** [comment]
@jeddy3 hey, thanks for your suggestions and complete answer!
I was also looking towards plugin :)
let me re-research that CLI thing, I'll be back
